module/zstd/zfs_zstd.c

```diff
@@ -198,6 +198,32 @@ static struct zstd_fallback_mem zstd_dctx_fallback;
 static struct zstd_pool *zstd_mempool_cctx;
 static struct zstd_pool *zstd_mempool_dctx;
 
+
+static void
+zstd_mempool_reap(struct zstd_pool *zstd_mempool)
+{
+	struct zstd_pool *pool;
+
+	if (!zstd_mempool) {
+		return;
+	}
+
+	/* free obsolete slots */
+	for (int i = 0; i < ZSTD_POOL_MAX; i++) {
+		pool = &zstd_mempool[i];
+		if (mutex_tryenter(&pool->barrier)) {
+			/* Free memory if unused object older than 2 minutes */
+			if (pool->mem && gethrestime_sec() > pool->timeout) {
```

Comment:
If the expectation is that many of these are NULL by the time you get here you may just get away with a quick check prior to locking in addition to the current code.

Reply:
mutex_tryenter is already such a basic check. mutex_tryenter checks first using "mutex_islocked" (linux implementation) which is a basic atomic_read check. no need to add another check. its already very lightweight

Reply:
Assuming there is 0 contention at play this will take (and release) all these locks, which is a significant avoidable cost especially on machines with high pool_count.

Reply:
I see this code always executes, even if there are no zstd datasets. This should be made as lightweight as possible. Apart from pre-checking pool->mem != NULL,  with stat collection from #11129 the entire thing could be short-circuited by inspecting total buffer count.

Reply:
what you describe increases the complexity and the total buffer count must be atomic again. this is not thread safe and needs extra locking just for checking the buffer count. not a good idea. keep it simple as possible and the current way is very lightweight.

Reply:
okay the buffer count is atomic. i added this extra check. but i can tell you that it will safe just a very few cpu cycles since the loop is small